### PR TITLE
MG transfer global coarsening: Avoid sending empty messages

### PR DESCRIPTION
**Pull request by kronbichler**
To make a test case of @marcfehling run on my MPI implementation, I had to ensure that there were no empty messages in addition to #16631, otherwise an MPI process would get stuck in the `BlackBoxFineDoFHandlerView::reinit()` function here: https://github.com/dealii/dealii/blob/d29226465a864be493f51fd051dfe811256ce4be/include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h#L972-L974 I believe the empty index sets in the `MPI::internal::ComputeIndexOwner` functions can happen when we have sparse index ranges as we might touch an `std::map` entry without filling it somewhere. I did not investigate that closer, but I think that all tests we have do not rely on empty messages, so I think we should expect that this is more robust anyway. @peterrum there might a place where we could filter out that earlier, but I can't say for sure where the empty index sets come from, so I would appreciate some help.

Furthermore, I can't give a test case because the one used by @marcfehling is rather large, and I could not understand what geometric configuration would be needed: Whenever I tried to reduce the complexity, the MPI deadlock would go away, so it seems my MPI implementation might simply have run out of buffers or something. @marcfehling would you mind testing your full case also with this patch, to ensure that nothing else happens here?

### CODE REVIEW

**Review by masterleinad — APPROVED** [review]

**Review by marcfehling — APPROVED** [review]
I tried your patch with the one testcase that caused all the issues related to #16615 and #16631. It worked fine with various counts of MPI processes (while oversubscribing). Everything went fine.

I will try some even bigger scenarios on the expanse supercomputer in the next few days. But we don't have to wait until then to merge this PR.

**Comment by peterrum** [comment]
> @peterrum there might a place where we could filter out that earlier, but I can't say for sure where the empty index sets come from, so I would appreciate some help.

This does not look right. A zero-size request means that there was no request (or there was a request that failed?)!? 

This should not happen. The question is whose side effect this is?

I guess the dictionary send these empty requesters here:

https://github.com/dealii/dealii/blob/da635b4ab900f4ba290fd1661df2bf41ac6ced6c/source/base/mpi_compute_index_owner_internal.cc#L636

Which are received by the index owners here:

https://github.com/dealii/dealii/blob/da635b4ab900f4ba290fd1661df2bf41ac6ced6c/source/base/mpi_compute_index_owner_internal.cc#L695-L699


This indicates that

https://github.com/dealii/dealii/blob/da635b4ab900f4ba290fd1661df2bf41ac6ced6c/source/base/mpi_compute_index_owner_internal.cc#L756-L761

is accidentally is adding an empty entry potentially at the wrong rank!?

I think it is worth to investigate this more to rule out that there is no bug. I cannot guarantee that I have time to debug this in the next days.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (bangerth) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:873]: `i.second` corresponds to an object of type `IndexSet`. Checking for `begin() == end()` is probably more easily understood if you said `i.second.is_empty()`.

Comment (bangerth) [reply include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:873]: Separately, I don't think this is sending an empty message. It perhaps sends an empty collection packed up into a string, but that's a non-zero length string.

**Comment by marcfehling** [comment]
Oh. I'm not sure right now whether we agreed that this patch does the right thing.

**Comment by peterrum** [comment]
> Oh. I'm not sure right now whether we agreed that this patch does the right thing.

No, we did not and should investigate the reason for the issue!

**Pull request merged**
